### DISCUSSION TITLE
Scroll item information with PgUp/PgDn on laptop

### DISCUSSION THREAD

**Discussion by gamolinf** [cataclysmbn/Cataclysm-BN#2064] [General]
Game version: 1943ad1

Operating system: Windows 10

Tiles or curses: UNDEAD PEOPLE (New versions)

Mods active: 

### Expected behavior
Can't use the keys PgUp/PgDown to scroll information about item on a laptop, if the keys PgUp/Pg/Down are combined with a small numeric keypad.
### Steps to reproduce the behavior
1) Use laptop
2) Take item with large discription, eg rm11b_sniper_rifle
3) Open inventory
4) Select item
5) Try scroll item information with PgUp/PgDn on the small numeric keypad

**Comment by Coolthulhu** [comment]
Probably related to the `Fn` key.
I'll have to check if there is a reliable way of getting keycodes with/without `Fn`, but I expect it to be very vendor-dependent and so probably unsolvable on our side.

**Comment by olanti-p** [comment]
DDA ticket for reference https://github.com/CleverRaven/Cataclysm-DDA/issues/28012

**Comment by Coolthulhu** [comment]
This is most likely `wontfix`, since the `Fn` key seems to be not just vendor specific, but not even detectable in software in any way.

What we CAN do:
* Add alternative binds for when `PgUp`/`PgDn` doesn't work. Possibly treat them as the primary ones, since keys outside QWERTY+numbers are less reliable. But most of the keyboard is bound already, might require context-specific binds. `[` and `]` seem like the safest choice, since they don't require shift and should be available on vast majority of keyboards.
* Add and maintain an alternative set of keybinds for laptops. This would probably be a ton of work, so unlikely to happen.
* Inform the user of the `Fn` key's potential effects on keybinds through tooltips. Especially the part where some laptops have "`Fn` inverted for most keys" as default option, usually changeable only in BIOS

What we CAN'T do:
* Make `PgUp`/`PgDn` "just work" on all laptops
* Have alternative keybinds for `Fn`+`PgUp`/`PgDn` without requiring the user to set them up first

**Comment by olanti-p** [comment]
I recently bought a laptop that has a similar problem. PgUp/PgDn are using same keys as 9 and 3 on the Numpad, and NumLock switches the behavior between producing a page key (when off) or a number key (when on).

The problem is, that works in all apps except for BN.

I managed to look into it with a debugger, and this seems to be a SDL issue.

Whenever I press Num3 with NumLock on, I get `SDL_KEYDOWN` event with this data:
![image](https://user-images.githubusercontent.com/60584843/189538247-725b2478-9b45-453f-a978-89ccc8cf29f1.png)

And with NumLock off, I get this:
![image](https://user-images.githubusercontent.com/60584843/189538275-5a6d5b3c-c77d-42c3-8294-e0480e902097.png)

As can be seen from above, the only meaningful difference is in `mod` value, with `4096` rather unsurprisingly indicating enabled/disabled NumLock key (`KMOD_NUM = 0x1000`).

There's no way we could infer the proper key from this without adding special cases for various laptop keyboard layouts.

It may be possible to work around this by somehow hacking into SDL message handling (iirc there was a way to get access to underlying Windows events), needs further research.

For reference, the testing was done in Visual Studio with SDL 2.20.0

**Comment by olanti-p** [comment]
Oh, also https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1802 made the item info scroll keys remappable, so this is not as much of an issue as it was. I've set them to F9/F10, where Brightness Up/Down Fn keys are, which makes it easy to remember which one is to scroll up and which one is to scroll down.
